README.md
# AutoLine开源平台
<!-- TOC -->

- [AutoLine开源平台](#autoline开源平台)
    - [什么是AutoLine开源平台](#什么是autoline开源平台)
    - [为什么要开发AutoLine](#为什么要开发autoline)
    - [特色](#特色)
    - [采用了哪些开源技术/框架](#采用了哪些开源技术框架)
    - [Mysql安装与配置](#mysql安装与配置)
    - [AutoLine下载及安装依赖](#autoline下载及安装依赖)
    - [配置文件说明](#配置文件说明)
    - [初始化及运行](#初始化及运行)
    - [外网访问运行方式](#外网访问运行方式)
    - [升级说明](#升级说明)
        - [一般版本升级](#一般版本升级)
        - [重大版本升级](#重大版本升级)
    - [如何支持不同的浏览器](#如何支持不同的浏览器)
    - [截图](#截图)
    - [支持与联系](#支持与联系)
    - [许可](#许可)
    - [公众号](#公众号)

<!-- /TOC -->

## 什么是AutoLine开源平台

AutoLine开源平台是一个开源自动化测试解决方案，基于RobotFramework进行二次开发，支持RobotFramework几乎所有的库。


用户简明教程： [简明教程](https://github.com/small99/AutoLine/blob/master/User%20Guide/User%20Guide/使用简明教程.md)

安装部署教程： [安装教程](https://github.com/small99/AutoLine/blob/master/User%20Guide/User%20Guide/安装部署教程.md)

知识星球交流圈子

![知识星球交流圈子](./app/static/images/quanzi.jpeg)

## 为什么要开发AutoLine

1. 为中小企业提供一个好用的自动化测试解决方案

2. 为广大软件测试从业人员提供一个基于开源框架的自动化测试平台开发示例

3. 业余写写，练习下敲代码的速度

## 特色

1. 支持Mac、Linux、windows平台

2. 支持Chrome、IE、Edge、Firefox、Safari、HtmlUnitDriver、PhantomJS、Appium、Requests HTTP

3. 支持与selenium-grid集成做分布式测试

等等

## 采用了哪些开源技术/框架

1. Python3

2. Flask

3. EasyUI

4. Chartjs

5. Robotframework

6. Requests

7. SQLAlchemy

8. Appium-Python-Client

9. Selenium3

等等其他及上述技术相关的插件或技术包

## Mysql安装与配置

先安装mysql数据库，使用utf-8字符编码新增autoline数据库

修改.env中关于数据库的配置，主要是：

```
DATABASE_URL=mysql+pymysql://root:123456@127.0.0.1/autoline
TRIGGER_DATABASE_URL=mysql+pymysql://root:123456@127.0.0.1/autoline
```

把上述两个配置中root:12345改为你的mysql的账户和密码
把127.0.0.1改为你的mysql的服务器地址即可

## AutoLine下载及安装依赖

下载AutoLine工程, 在AutoLine根目录下有requirements.txt文件，该文件包含了所有依赖的包，安装好Python3并确保Python3的pip可用，使用以下命令安装AutoLinie依赖库

> pip install -r requirements.txt

等待安装完成

## 配置文件说明
在AutoLine目录下的.env为所有关键的配置项

```INI
# 用于定义运行环境
# production - 生产环境
# development - 开发环境 
AUTOBEAT_CONFIG=production

# 服务器端启动秘钥，这里随意输入一串字符
SECRET_KEY=FJWEOIO138989340SNNFISLDNFIOWFNSNALFKNSDLKF

# SMTP账号
MAIL_SERVER=
MAIL_PORT=587
MAIL_USERNAME=
MAIL_PASSWORD=25
FLASKY_MAIL_SUBJECT_PREFIX=[AutoLine]
FLASKY_MAIL_SENDER=AutoLine Admin <autoline@126.com>

# 默认admin账号, 默认密码123456
FLASKY_ADMIN=autoline@126.com

# 数据库连接字符串
DATABASE_URL=mysql+pymysql://root:123456@127.0.0.1/autoline

# 调度连接字符串
TRIGGER_DATABASE_URL=mysql+pymysql://root:123456@127.0.0.1/autoline
```

## 初始化及运行

在配置好上述配置后如何运行AutoLine？

一. 初始化数据库

> python manage.py db init

> python manage.py db migrate

> python manage.py db upgrade

二. 初始化数据

> python manage.py deploy

三. 内网运行方式

> python manage.py runserver

打开浏览器访问：http://127.0.0.1:5000即可

## 外网访问运行方式

使用下面的命令即可

> python manage.py runserver -h 0.0.0.0 -p 8000

-h 指定ip（ip指定为0.0.0.0即可自定绑定本机IP）    

-p 指定端口号

这时你就可以用在其他机器通过： http://ip:端口     来访问平台了

![首页](./app/static/images/demo/index.png)

使用默认账号： autoline@126.com/123456 登录如图：

![start](./app/static/images/demo/start.png)  

## 升级说明

### 一般版本升级

如果你使用的是git，直接用下列命令升级即可

> git pull

实现源码升级

否则需要从github下载源码进行覆盖升级！！！

### 重大版本升级

主要涉及了数据模型的修改出了代码进行升级外，还需要重新升级数据模型，源码升级同上，下面为数据库模型升级：

1. 删除本地的migrations目录

2. 清空数据库表alembic_version中所有内容

3. 按以下步骤在命令行中使用下列命令升级数据库模型

> python manage.py db init

> python manage.py db migrate

> python manage.py db upgrade

4. 升级完成启动服务

> python manage.py runserver

## 如何支持不同的浏览器

下载最新的各浏览器对应的最新驱动，放在bin目录下即可

firefox驱动：https://github.com/mozilla/geckodriver/releases

chrome驱动：https://chromedriver.storage.googleapis.com/index.html

Microsofe Edge驱动：https://developer.microsoft.com/en-us/microsoft-edge/tools/webdriver/

Safari驱动：https://webkit.org/blog/6900/webdriver-support-in-safari-10/

PhantomJS驱动：http://phantomjs.org/releases.html

HtmlUnit驱动：https://sourceforge.net/projects/htmlunit/files/htmlunit/

## 截图
调试模式
![调试模式](./app/static/images/demo/debug_mode.png)

产品管理
![产品管理](./app/static/images/demo/product.png)

项目管理
![项目管理](./app/static/images/demo/project.png)

调度管理
![调度管理](./app/static/images/demo/scheduler_manage.png)


运行日志
![查看任务](./app/static/images/demo/run_logs.png)
![查看任务](./app/static/images/demo/run_logs_detail.png)

查看任务
![查看任务](./app/static/images/demo/task.png)

测试报告
![查看报告](./app/static/images/demo/run_logs_detail.png)

关键字帮助文档
![帮助手册](./app/static/images/demo/help.png)

## 支持与联系

公众号：开源优测  

email：lymking#foxmail.com    

知识星球：开源优测  

关于开源测试技术相关文章首发公众号。

## 许可

AutoLine基于Apache License 2.0协议发布

## 公众号

扫一扫关注我的公众号，获取第一手文档资料：

![开源优测](./app/static/images/deeptest.jpg)


history.md
#### 2018-07-03 v1.1.3
1. 将关键字表中的keyword字段修改为128长度，以解决不同mysql版本兼容性问题

#### 2018-06-04 v1.1.2
1. 将连接字符串驱动从pymysql修改为mysqlconnector

#### 2018-06-01 v1.1.1
1. 注释自定义关键字套件右击"管理关键字"功能
2. 新增删除自定义关键字功能

#### 2018-05-31 v1.1.0
1. 重大版本更新，修订了数据库模型，新增了自定义关键字支持
2. 修复用户反馈的缺陷

#### 2018-05-29 v1.0.18
1. 本版本为强制更新版本
2. 修复用户反馈默认的关键字错误的缺陷
3. 优化步骤中关键字列表，改为根据项目类型来显示合适的关键字列表

#### 2018-05-29 v1.0.17
1. 修订用户反馈的关键字列表问题
2. 注释可能导致跨平台兼容性的问题

#### 2018-05-28 v1.0.16
1. 新增项目运行项目时先判断是否有关键字步骤存在，若有则运行，若无则提示用户创建步骤

#### 2018-05-28 v1.0.15
1. 创建项目时，cron表达式设置为必填项，并新增cron表达式格式说明及示例

#### 2018-05-24 v1.0.14
1. 新增mysql连接字符串utf8编码（运行autoline会有warning出现，请忽略该warning即可）

#### 2018-05-23 v1.0.13
1. 新增同一项目下对象重复的限制


#### 2018-05-21 v1.0.11
1. 修复windows下查看日志转码问题

#### 2018-05-21 v1.0.10
1. 修订任务状态显示异常问题
2. 新增异常任务状态显示
3. 修订报告显示样式
4. 修订启动时版本检测输出

#### 2018-05-21 v1.0.9
1. 新增项目类型能力，支持web、app、http三类（对应web ui、app ui以及http接口测试项目）

#### 2018-05-21 v1.0.8
1. 修订详细报告显示不完整的缺陷

#### 2018-05-19 v1.0.7
1. 根据os类型执行不同的进程创建参数

#### 2018-05-19 v1.0.6
1. 修订报告显示错误
2. 修订启动RF进程时的参数
3. 修订启动时版本检测地址，改为从码云上获取版本信息

#### 2018-05-19 v1.0.5
1. 修订windows下乱码问题
2. 修复截图显示问题

#### 2018-05-19 v1.0.3
修复在windows报：[WinError 2] 系统找不到指定文件的缺陷

#### 2018-05-19
1. 新增启动AutoLine时，代码版本检测
2. 新增更新列表清单文件
3. 新增调试模式
4. 丰富测试报告内容
5. 丰富调度管理：手工管理模式
6. 修订文档及截图
7. 新增运行日志显示
8. 修复用户反馈的细节缺陷

#### 2018-05-16
1. 新增对象创建时规则验证，防止逗比式的乱命名
2. 修复选择第一个对象时，填充输入框和选择值不一致问题

#### 2018-05-15
1. 修订文档及相关图片
2. 修订用户反馈的bug
3. 注释一些不必要的代码

#### 2018-05-14
1. 修订在Windows下安装部署时的一些问题
2. 新增windows 64位firefox和chrome驱动
3. 修复一些反馈的bug
4. 新增pymysql依赖

#### 2018-05-11
1. 将AutoLine开源平台发布至github开源

config.py
# -*- coding: utf-8 -*-

__author__ = "苦叶子"

"""

公众号: 开源优测

Email: lymking@foxmail.com

"""

import os
import random
import codecs

basedir = os.path.abspath(os.path.dirname(__file__))

choice = ("QWERTYUIOPASDFGHJKLMNBVCXZqazxswedcvfrtgbnhyujmkiolp1234567890!@#$%^&*")


class Config:
    if os.path.exists('.env'):
        print('Import environment from .env')
        for line in codecs.open('.env'):
            var = line.strip().split('=')
            if len(var) == 2:
                os.environ[var[0]] = var[1]
                #print(var[0] + "=" + var[1])

    SECRET_KEY = os.environ.get('SECRET_KEY')# or random.choices(choice, k=16)
    MAIL_SERVER = os.environ.get('MAIL_SERVER', 'smtp.126.com')
    MAIL_PORT = int(os.environ.get('MAIL_PORT', '587'))
    MAIL_USE_TLS = os.environ.get('MAIL_USE_TLS', 'true').lower() in ['true', 'on', '1']
    MAIL_USERNAME = os.environ.get('MAIL_USERNAME')
    MAIL_PASSWORD = os.environ.get('MAIL_PASSWORD')
    FLASKY_MAIL_SUBJECT_PREFIX = os.environ.get('FLASKY_MAIL_SUBJECT_PREFIX')
    FLASKY_MAIL_SENDER = os.environ.get('FLASKY_MAIL_SENDER')
    FLASKY_ADMIN = os.environ.get('FLASKY_ADMIN')
    SSL_REDIRECT = False
    SQLALCHEMY_TRACK_MODIFICATIONS = False
    SQLALCHEMY_RECORD_QUERIES = True
    FLASKY_POSTS_PER_PAGE = 20
    FLASKY_FOLLOWERS_PER_PAGE = 50
    FLASKY_COMMENTS_PER_PAGE = 30
    FLASKY_SLOW_DB_QUERY_TIME = 0.5

    # 自定义参数
    AUTO_LOGS = os.environ.get('AUTO_LOGS')
    AUTO_REPORT = os.environ.get('AUTO_REPORT')

    RUNNERS = []
    TRIGGER = None

    @staticmethod
    def init_app(app):
        pass


class DevelopmentConfig(Config):
    DEBUG = True

    SQLALCHEMY_DATABASE_URI = 'sqlite:///' + os.path.join(basedir, 'AutoLine-dev.sqlite') #os.environ.get('DEV_DATABASE_URL')
    TRIGGER_DATABASE_URL = 'sqlite:///' + os.path.join(basedir, 'AutoLine-dev.sqlite')


class ProductionConfig(Config):
    DEBUG = True
    SQLALCHEMY_DATABASE_URI = os.environ.get('DATABASE_URL') + "?charset=utf8" # or 'sqlite:///' + os.path.join(basedir, 'AutoLine.sqlite')
    #SQLALCHEMY_ECHO = True
    TRIGGER_DATABASE_URL = os.environ.get('TRIGGER_DATABASE_URL')


    @classmethod
    def init_app(cls, app):
        Config.init_app(app)

        # 发送初始化错误信息给管理员
        import logging
        from logging.handlers import SMTPHandler
        credentials = None
        secure = None
        if getattr(cls, 'MAIL_USERNAME', None) is not None:
            credentials = (cls.MAIL_USERNAME, cls.MAIL_PASSWORD)
            if getattr(cls, 'MAIL_USE_TLS', None):
                secure = ()

        mail_handler = SMTPHandler(
            mailhost=(cls.MAIL_SERVER, cls.MAIL_PORT),
            fromaddr=cls.FLASKY_MAIL_SENDER,
            toaddrs=[cls.FLASKY_ADMIN],
            subject=cls.FLASKY_MAIL_SUBJECT_PREFIX + ' AutoLine Startup Error',
            credentials=credentials,
            secure=secure)

        mail_handler.setLevel(logging.ERROR)
        app.logger.addHandler(mail_handler)


config = {
    "development": DevelopmentConfig,
    "production": ProductionConfig,

    "default": DevelopmentConfig
}


# 关键字自定义配置
USER_KEYS = {
    "web": ["BuiltIn", "Collections", "String", "DateTime", "Screenshot", "SeleniumLibrary"],
    "app": ["BuiltIn", "Collections", "String", "DateTime", "Screenshot", "AppiumLibrary"],
    "http": ["BuiltIn", "Collections", "String", "DateTime", "RequestsLibrary"],
    "all": ["BuiltIn", "Collections", "String", "Screenshot", "DateTime",
            "SeleniumLibrary", "AppiumLibrary", "RequestsLibrary",
            "Process", "Telnet", "OperatingSystem"]
}
